jcl/src/java.base/share/classes/openj9/internal/foreign/abi/LayoutStrPreprocessor.java

```diff
@@ -209,38 +215,153 @@ private static StringBuilder preprocessLayout(MemoryLayout targetLayout, boolean
+	/* Encode the types in the union layout to a symbol string to simplify the further processing in native. */
+	private static StringBuilder encodeUnionLayoutStr(UnionLayout unionLayout, StringBuilder targetLayoutStr, boolean isDownCall) {
+		long unionByteSize = unionLayout.byteSize();
+		StringBuilder elementLayoutStrs = new StringBuilder("");
+		int byteAlignment = (int)unionLayout.byteAlignment();
+		ValueLayout sequenceElement = null;
+
+		/* The idea is to select the primitve type with the biggest alignement of the union
+		 * (obtained via unionLayout.byteAlignment()) to be laid out as a sequence which is
+		 * correctly mappped onto the given GPRs/FPRs on the hardware.
+		 */
+		switch (byteAlignment) {
+		case 1: /* 1 Byte. */
+			sequenceElement = JAVA_BYTE;
+			break;
+		case 2: /* 2 Bytes. */
+			sequenceElement = JAVA_SHORT;
+			break;
+		case 4: /* 4 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_INT;
+			} else {
+				sequenceElement = JAVA_FLOAT;
 			}
-			targetLayoutString.append('[').append(elementLayoutStrs).append(']');
+			break;
+		case 8: /* 8 Bytes. */
+			if (validatePrimTypesOfUnionForGPR(unionLayout, byteAlignment)) {
+				sequenceElement = JAVA_LONG;
+			} else {
+				sequenceElement = JAVA_DOUBLE;
+			}
+			break;
+		default:
+			throw new IllegalArgumentException("Invalid alignment of the union layout: " + unionLayout);
 		}
 
-		return targetLayoutString;
+		long elementByteSize = sequenceElement.byteSize();
+		long sequenceCount = (0 == (unionByteSize % elementByteSize)) ?
+				(unionByteSize / elementByteSize) : (unionByteSize / elementByteSize + 1);
+
+		/* There is no need to create a sequence layout for a single element. */
+		MemoryLayout sequenceLayout = (sequenceCount > 1) ?
+				MemoryLayout.sequenceLayout(sequenceCount, sequenceElement) : sequenceElement;
+		elementLayoutStrs.append(preprocessLayout(sequenceLayout, isDownCall));
+
+		/* Prefix "#" to denote the start of this layout string in the case of downcall. */
+		if (isDownCall) {
+			targetLayoutStr.append('#').append(1);
+		}
+		targetLayoutStr.append('[').append(elementLayoutStrs).append(']');
+
+		return targetLayoutStr;
 	}
 
+	/* Check whether the specified 4/8-byte aligned group layout (struct/union) layout contains
+	 * non-floating-point primitves so as to determine whether the elements of union can be safely
+	 * mapped to GPRs.
+	 */
+	private static boolean validatePrimTypesOfUnionForGPR(GroupLayout groupLayout, int byteAlignment) {
+		List<MemoryLayout> elementLayoutList = groupLayout.memberLayouts();
+		boolean arePrimTypesForGPR = false;
+		for (MemoryLayout elementLayout : elementLayoutList) {
+			if (((elementLayout instanceof ValueLayout elemValueLayout)
+				&& (validatePrimTypeForGPR(elemValueLayout, byteAlignment)))
+				|| ((elementLayout instanceof GroupLayout elemGroupLayout)
+				&& validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))
```

Comment:
formating, `&& validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))` should be indented to denote it is part of the `||`

Reply:
actually its more clear like this

```
if (((elementLayout instanceof ValueLayout elemValueLayout) && (validatePrimTypeForGPR(elemValueLayout, byteAlignment)))
    || ((elementLayout instanceof GroupLayout elemGroupLayout) && validatePrimTypesOfUnionForGPR(elemGroupLayout, byteAlignment))
) {
	
}
```

Reply:
Agreed and updated against the suggestion above.